lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -34,421 +25,546 @@
+    auto numNodes = graph.size();
+    SmallVector<bool> onStack(numNodes, false);
+    SmallVector<unsigned> dfsStack;
+
+    auto hasCycle = [&](unsigned rootNode, DenseSet<unsigned> &visited,
+                        bool recordPortPaths = false) {
+      if (visited.contains(rootNode))
+        return success();
+      dfsStack.push_back(rootNode);
+
+      while (!dfsStack.empty()) {
+        auto currentNode = dfsStack.back();
+
+        if (!visited.contains(currentNode)) {
+          visited.insert(currentNode);
+          onStack[currentNode] = true;
+          LLVM_DEBUG(llvm::dbgs()
+                     << "\n visiting :"
+                     << reachingDefGraph[currentNode].first.getValue().getType()
+                     << reachingDefGraph[currentNode].first.getValue() << ","
+                     << reachingDefGraph[currentNode].first.getFieldID() << "\n"
+                     << getName(reachingDefGraph[currentNode].first));
+
+          FieldRef currentF = reachingDefGraph[currentNode].first;
+          if (recordPortPaths && currentNode != rootNode) {
+            if (isa<mlir::BlockArgument>(currentF.getValue()))
+              portPaths[reachingDefGraph[rootNode].first].insert(currentF);
+            // Even if the current node is not a port, there can be RWProbes of
+            // the current node at the port.
+            addToPortPathsIfRWProbe(
+                currentNode, portPaths[reachingDefGraph[rootNode].first]);
+          }
+        } else {
+          onStack[currentNode] = false;
+          dfsStack.pop_back();
+        }
+
+        for (auto neighbor : graph[currentNode].second) {
+          if (!visited.contains(neighbor)) {
+            dfsStack.push_back(neighbor);
+          } else if (onStack[neighbor]) {
+            // Cycle found !!
+            SmallVector<FieldRef, 16> path;
+            auto loopNode = neighbor;
+            // Construct the cyclic path.
+            do {
+              SmallVector<unsigned>::iterator it =
+                  llvm::find_if(reachingDefGraph[loopNode].second,
+                                [&](unsigned node) { return onStack[node]; });
+              if (it == reachingDefGraph[loopNode].second.end())
+                break;
+
+              path.push_back(reachingDefGraph[loopNode].first);
+              loopNode = *it;
+            } while (loopNode != neighbor);
+
+            reportLoopFound(path, reachingDefGraph[neighbor].first.getLoc());
+            return failure();
+          }
+        }
+      }
+      return success();
     };
-    auto errorDiag = mlir::emitError(
-        module.getLoc(), "detected combinational cycle in a FIRRTL module");
 
-    SmallVector<Value, 16> path;
-    path.push_back(childVal);
-    visiting.popUntilVal(
-        childVal, [&](Value visitingVal) { path.push_back(visitingVal); });
-    assert(path.back() == childVal);
-    path.pop_back();
+    DenseSet<unsigned> visited;
```

Comment:
Noticed this is a set while `onStack` is a vector including all nodes eagerly in its allocation/size.  Since `visited` is expected to eventually include all nodes (right?) I'm wondering what the thinking is here-- presumably any given DFS walk is expected to have a smaller set of things it encounters compared to populating the entire set (visited), but dunno just curious what motivated that.

(Is this a set to make `clear()` more efficient, maybe?)

Reply:
I don't exactly remember the reasoning, but the vector `onStack` is required to ensure deterministic error reporting.